### ISSUE TITLE
Server-side Rendering

### ISSUE THREAD

**Issue by wKovacs64** [tizmagik/react-csjs#24]
Isomorphic/universal React is a thing and react-csjs should probably support it in some capacity. Out of the box, it won't work in that environment because it inserts styles into the global `document` object in the component lifecycle methods, which doesn't exist server-side.

We should probably either solve it internally (not sure how or if that's even possible), or provide documentation on how to do it externally. I have a first-stab ~demo~ (edit: now defunct) of solving it externally, but I'm not sure if it's the right approach (I've never personally done server-side rendering, so I'm treading deep water here).

Thoughts, @tizmagik ?

**Comment by tizmagik** [comment]
Yes, totally! Nice work with the demo, it gives a good indication of where the boundaries are that need to move to make `react-csjs` isomorphic.

What I'd love to see is `insert-style.js` could sniff if `document` is defined, and if not, it could instead act as a singleton, collecting all the styles that get passed in. In fact, the `csjs` object that is used internally could be treated as the singleton, which would actually take care of #2 as well. So that something like the following is possible:

``` jsx
import React from 'react';
import { renderToString } from 'react-dom/server';
import express from 'express';
import allStyles from 'react-csjs/server'; // not sure about this interface
import App from './App';

const app = express();

app.use((req, res) => {
  const renderedApp = renderToString(<App />);
  const criticalStyles = allStyles();

  res.send(`
    <html>
      <head><style>${criticalStyles}</style></head>
      <body>
        <main id="root">${renderedApp}</main>
        <script src="/main.js"></script>
      </body>
    </html>
  `);
});
```

So basically, `allStyles()` (we should think of a better name/interface) would return the singleton-collected styles that the app made during `renderToString()`. This would be, by definition, the "critical CSS" necessary to render just the server-side payload. Then on the client there might be an extra call or something to remove the SSR rendered CSS because at that point, `react-csjs` would inject the styles needed as each component mounts on the client.

What do you think @wKovacs64  ? @drcmda would like to get your thoughts here as well.

**Comment by tizmagik** [comment]
@wKovacs64 for the time being, I think it might make sense to add some documentation for a recipe on how to solve it externally, as I imagine it might take us some time to get proper isomorphic support just right. Happy to accept a PR for that.

**Comment by wKovacs64** [comment]
Yeah, that sounds good. I can try to throw a PR together with a note in the documentation about it for the time being. Do we want to move my demo into the `example` folder (separate from the existing one - we can make a `basic` subdirectory for that one or something)? Or link to it? If we do include/reference it, someone more experienced may want to see if they can simplify my webpack configs; it got a little weird in there. 😉

**Comment by tizmagik** [comment]
I think just a note for now would be good. You can link to your demo externally but I rather keep the example app simple. 

Once we implement working isomorphic then we can update the example app to show isomorphic usage.
